readthedocs/notifications/decorators.py

```diff
@@ -0,0 +1,35 @@
+# -*- coding: utf-8 -*-
+
+import logging
+
+from django.db.models import Q
+
+from readthedocs.projects.models import Project
+from readthedocs.projects.notifications import DeprecatedWebhookEndpointNotification
+
+log = logging.getLogger(__name__)
+
+
+def notify_deprecated_endpoint(function):
+    """
+    Decorator to notify owners that the endpoint is DEPRECATED.
+    """
+    def wrap(request, *args, project_id_or_slug=None, **kwargs):
+        try:
+            project = Project.objects.get(
+                Q(pk=project_id_or_slug) | Q(slug=project_id_or_slug),
+            )
+        except (Project.DoesNotExist, ValueError):
+            log.info('Project not found: slug=%s', project_id_or_slug)
+
+        # Send one notification to each owner of the project
+        for user in project.users.all():
+            notification = DeprecatedWebhookEndpointNotification(
+                project,
+                request,
+                user,
+            )
+            notification.send()
```

Comment:
By calling `.send()` here we will be sending an email notification each time the endpoint is hit and that's not right.

Since we only have `email` (send email immediately) and `site` (make a persistent notification) backends I'm thinking on creating a new called `delayed_email` which would be a mixture between these two: save the notification in the storage to be sent as an email later.

I think we need here:

* create notifications on each endpoint hit
* deduplicate them automatically when created
* store them somewhere
* recover the notification stored (and filter by type `deprecated_webhook_endpoint`) from a Celery task
* send them and remove them from the storage

cc @agjohnson

Reply:
Either our code or message-extends already does this for us. It de-duplicates messages. I'd have to dig up where this is, but a third backend shouldn't be necessary

Reply:
Yes, but it's a custom storage https://github.com/rtfd/readthedocs.org/blob/5fb61ac617741dc3c9adcacb64d5e7154deedab2/readthedocs/notifications/storages.py#L28

Since we are not calling `storage.add` when using the `EmailNotification` that code is never called, from what I understand. Right?

Reply:
What I'm proposing is to create a custom backend but use that same storage that already de-duplicate the messages.

Reply:
Yeah, our custom fallback storage backend:
https://github.com/rtfd/readthedocs.org/blob/master/readthedocs/notifications/storages.py#L20-L80